File: dsa.js
let player;
let currentUser = null;
const users = {}; // In-memory user storage

// Sample videos that are publicly embeddable
const sampleVideos = [
    { id : 'xuP4g7IDgDM', title: 'Beautiful Sunset' },
    { id : 'nqye02H_H6I', title: 'River video' },
    { id : '9iDXWx7GtZQ', title: 'Soothing nature' },
    { id : '-oOoTIuoL8M', title: 'Mountain'},
    { id : 'NSAOrGb9orM', title: 'Beautiful  Landscape' },


];

class VideoStack {
  constructor() {
    this.videos = [];
    this.capacity = 10;
    this.currentIndex = 0;
  }
  push(video) {
    if (this.videos.length >= this.capacity) return;
    this.videos.push(video);
    updateStackDisplay();
    updateStackInfo();
  }
  pop() {
    if (this.videos.length === 0) return null;
    const v = this.videos.pop();
    updateStackDisplay();
    updateStackInfo();
    return v;
  }
  peek() {
    return this.videos.length === 0 ? null : this.videos[this.videos.length - 1];
  }
  size() {
    return this.videos.length;
  }
  getCurrent() {
    if (this.videos.length === 0) return null;
    return this.videos[this.currentIndex % this.videos.length];
  }
  next() {
    if (this.videos.length === 0) return null;
    this.currentIndex = (this.currentIndex + 1) % this.videos.length;
    return this.getCurrent();
  }
}

const stack = new VideoStack();

function onYouTubeIframeAPIReady() {
  player = new YT.Player('player', {
    height: '300',
    width: '100%',
    videoId: '',
    playerVars: {
      autoplay: 0,
      controls: 1,
      modestbranding: 1,
      rel: 0,
      enablejsapi: 1,
      playsinline: 1,
    },
    events: {
      onReady: onPlayerReady,
      onStateChange: onPlayerStateChange,
      onError: onPlayerError,
    },
  });
}

let skipTimer = null;

function onPlayerReady(event) {
  if (stack.size() > 0) {
    playCurrentVideo();
  }
}

function onPlayerStateChange(event) {
  // When video starts playing, start 5 sec skip timer
  if (event.data == YT.PlayerState.PLAYING) {
    if (skipTimer) clearTimeout(skipTimer);
    skipTimer = setTimeout(() => {
      // Skip to next video after 5 seconds
      playNextVideo();
    }, 5000);
  } else {
    if (skipTimer) {
      clearTimeout(skipTimer);
      skipTimer = null;
    }
  }
}

function onPlayerError(event) {
  console.warn('Error in playback. Skipping video. Error code:', event.data);
  playNextVideo();
}

function playCurrentVideo() {
  const current = stack.getCurrent();
  if (!current) return;
  player.loadVideoById(current.id);
  updateStackInfo();
}

function playNextVideo() {
  const next = stack.next();
  if (!next) {
    player.stopVideo();
    return;
  }
  player.loadVideoById(next.id);
  updateStackInfo();
}

function updateStackDisplay() {
  const stackItems = document.getElementById('stackItems');
  stackItems.innerHTML = '';
  if (stack.size() === 0) {
    stackItems.innerHTML = '<p>Stack is empty</p>';
    return;
  }
  for (let i = stack.videos.length - 1; i >= 0; i--) {
    const v = stack.videos[i];
    const item = document.createElement('div');
    item.className = 'stack-item';
    item.textContent = v.title;
    stackItems.appendChild(item);
  }
}

function updateStackInfo() {
  const stackSize = document.getElementById('stackSize');
  const topElement = document.getElementById('topElement');
  stackSize.textContent = stack.size();
  const current = stack.getCurrent();
  topElement.textContent = current ? current.title : 'None';
}

document.addEventListener('DOMContentLoaded', () => {
  const loginModal = document.getElementById('loginModal');
  const signupModal = document.getElementById('signupModal');
  const authBtn = document.getElementById('authBtn');
  const loginClose = loginModal.querySelector('.close');
  const signupClose = signupModal.querySelector('.close');
  const switchToSignup = document.getElementById('switchToSignup');
  const switchToLogin = document.getElementById('switchToLogin');
  const loginForm = document.getElementById('authForm');
  const signupForm = document.getElementById('signupForm');
  const mainContent = document.getElementById('mainContent');

  function showLogin() {
    loginModal.style.display = 'block';
  }
  function showSignup() {
    signupModal.style.display = 'block';
  }
  function hideModals() {
    loginModal.style.display = 'none';
    signupModal.style.display = 'none';
  }

  authBtn.addEventListener('click', () => {
    if (!currentUser) {
      showLogin();
    }
  });
  loginClose.addEventListener('click', () => {
    loginModal.style.display = 'none';
  });
  signupClose.addEventListener('click', () => {
    signupModal.style.display = 'none';
  });

  switchToSignup.addEventListener('click', (e) => {
    e.preventDefault();
    hideModals();
    showSignup();
  });
  switchToLogin.addEventListener('click', (e) => {
    e.preventDefault();
    hideModals();
    showLogin();
  });

  loginForm.addEventListener('submit', (e) => {
    e.preventDefault();
    const username = loginForm.username.value;
    const password = loginForm.password.value;
    if (users[username] && users[username] === password) {
      currentUser = username;
      updateAuthUI();
      hideModals();
      loginForm.reset();
      mainContent.style.display = 'grid';
      if (stack.size() > 0) playCurrentVideo();
    } else {
      alert('Invalid username or password!');
    }
  });

  signupForm.addEventListener('submit', (e) => {
    e.preventDefault();
    const username = signupForm.signupUsername.value;
    const password = signupForm.signupPassword.value;
    const confirmPassword = signupForm.confirmPassword.value;
    if (password !== confirmPassword) {
      alert('Passwords do not match!');
      return;
    }
    if (users[username]) {
      alert('Username already exists!');
      return;
    }
    users[username] = password;
    currentUser = username;
    updateAuthUI();
    hideModals();
    signupForm.reset();
    mainContent.style.display = 'grid';

    // Add demo videos on signup
    sampleVideos.forEach((vid) => stack.push(vid));
    if (stack.size() > 0) playCurrentVideo();
  });

  function updateAuthUI() {
    if (currentUser) {
      authBtn.innerHTML = `Logged in as ${currentUser}`;
      authBtn.disabled = true;
    } else {
      authBtn.innerHTML = 'Login / Sign Up';
      authBtn.disabled = false;
      mainContent.style.display = 'none';
    }
  }

  // Buttons to manually control video playback and stack
  document.getElementById('playBtn').onclick = () => {
    const current = stack.getCurrent();
    if (current) player.playVideo();
  };
  document.getElementById('pauseBtn').onclick = () => {
    player.pauseVideo();
  };
  document.getElementById('pushBtn').onclick = () => {
    // Push a random video from sample list
    const v = sampleVideos[Math.floor(Math.random() * sampleVideos.length)];
    stack.push(v);
    playCurrentVideo();
  };
  document.getElementById('popBtn').onclick = () => {
    stack.pop();
    const current = stack.getCurrent();
    if (current) playCurrentVideo();
    else player.stopVideo();
  };
  document.getElementById('traverseBtn').onclick = () => {
    const current = stack.getCurrent();
    if (current) playCurrentVideo();
  };

  updateAuthUI();
});

File: dsa.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define MAX_VIDEOS 10
#define SKIP_INTERVAL 5
#define MAX_USERS 50
#define CRED_FILE "credentials.dat"

typedef struct {
  char path_or_id[100];
  char title[100];
  int is_mp4;
} Video;

typedef struct {
  Video videos[MAX_VIDEOS];
  int top;
} VideoStack;

typedef struct {
  char username[50];
  char password[50];
} User;

void push(VideoStack *stack, Video video) {
  if (stack->top >= MAX_VIDEOS - 1) {
    printf("Stack overflow: Cannot push \"%s\"\n", video.title);
    return;
  }
  stack->videos[++(stack->top)] = video;
  printf("Pushed: \"%s\"\n", video.title);
}

Video pop(VideoStack *stack) {
  Video empty = {"", "", 0};
  if (stack->top < 0) {
    printf("Stack underflow: Cannot pop from empty stack\n");
    return empty;
  }
  Video video = stack->videos[(stack->top)--];
  printf("Popped: \"%s\"\n", video.title);
  return video;
}

Video peek(VideoStack *stack) {
  Video empty = {"", "", 0};
  if (stack->top < 0) return empty;
  return stack->videos[stack->top];
}

void traverse(VideoStack *stack) {
  if (stack->top < 0) {
    printf("Stack is empty\n");
    return;
  }
  printf("Stack Contents (Top to Bottom):\n");
  for (int i = stack->top; i >= 0; --i) {
    Video v = stack->videos[i];
    if (v.is_mp4)
      printf(" - %s (Local MP4: %s)\n", v.title, v.path_or_id);
    else
      printf(" - %s (YouTube ID: %s)\n", v.title, v.path_or_id);
  }
}

void playAndSkip(VideoStack *stack) {
  if (stack->top < 0) {
    printf("No video to play\n");
    return;
  }

  int i = stack->top;
  while (1) { // Loop forever through the stack
    Video current = stack->videos[i];
    printf("\nPlaying: %s\n", current.title);
    if (current.is_mp4)
      printf("Source: %s (Local MP4 File)\n", current.path_or_id);
    else
      printf("YouTube ID: %s\n", current.path_or_id);

    int currentTime = 0;
    while (currentTime < 25) { // play up to 25 seconds max for demo
      printf("Time: %ds - %s\n", currentTime, current.title);
      currentTime += SKIP_INTERVAL;
      sleep(1);
      printf("Skipped %d seconds!\n", SKIP_INTERVAL);
    }
    printf("Finished playing: %s\n", current.title);

    i--;
    if (i < 0) i = stack->top; // circle back to top (loop)
  }
}

// --- User credential storage ---

int saveUserCredential(User user) {
  FILE *fp = fopen(CRED_FILE, "ab");
  if (!fp) {
    perror("Error opening credential file");
    return 0;
  }
  fwrite(&user, sizeof(User), 1, fp);
  fclose(fp);
  return 1;
}

// Check if user exists and password matches
int validateUser(char *username, char *password) {
  FILE *fp = fopen(CRED_FILE, "rb");
  if (!fp) return 0; // no credentials file means no users stored
  User user;
  while (fread(&user, sizeof(User), 1, fp)) {
    if (strcmp(user.username, username) == 0 && strcmp(user.password, password) == 0) {
      fclose(fp);
      return 1; // user validated
    }
  }
  fclose(fp);
  return 0; // no matching user
}

// Check if username already exists
int userExists(char *username) {
  FILE *fp = fopen(CRED_FILE, "rb");
  if (!fp) return 0;
  User user;
  while (fread(&user, sizeof(User), 1, fp)) {
    if (strcmp(user.username, username) == 0) {
      fclose(fp);
      return 1;
    }
  }
  fclose(fp);
  return 0;
}

int main() {
  Video videos[] = {

    {"xuP4g7IDgDM", "Beautiful sunset", 0},
    {"nqye02H_H6I", "River video", 0},
    {"9iDXWx7GtZQ", "Soothing nature", 0},
    {"NSAOrGb9orM", "Beautiful landscape", 0},
    {"-oOoTIuoL8M", "Mountain", 0}
  };

  int sampleCount = sizeof(videos)/sizeof(videos[0]);

  VideoStack stack;
  stack.top = -1;

  // Add videos to stack
  for (int i = 0; i < sampleCount; ++i) {
    push(&stack, videos[i]);
  }

  traverse(&stack);

  // Play videos skipping every 5 seconds in an infinite loop
  playAndSkip(&stack);

  return 0;
}
